shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -132,19 +140,56 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
   public boolean finishComposingText() {
     boolean result = super.finishComposingText();
 
-    if (Build.VERSION.SDK_INT >= 21) {
-      // Update the keyboard with a reset/empty composing region. Critical on
-      // Samsung keyboards to prevent punctuation duplication.
-      CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
-      builder.setComposingText(-1, "");
-      CursorAnchorInfo anchorInfo = builder.build();
-      mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+    // Apply Samsung hacks. Samsung caches composing region data strangely, causing text duplication.
+    if (mIsSamsung) {
+      if (Build.VERSION.SDK_INT >= 21) {
+        // Update the keyboard with a reset/empty composing region. Critical on
+        // Samsung keyboards to prevent punctuation duplication.
+        CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
+        builder.setComposingText(-1, "");
+        CursorAnchorInfo anchorInfo = builder.build();
+        mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+      }
+      // TODO(garyq): There is still a duplication case that comes from hiding+showing the keyboard.
+      // The exact behavior to cause it has so far been hard to pinpoint and it happens far more
+      // rarely than the original bug.
+
+      // Temporarily indicate to the IME that the composing region selection should be reset.
+      // The correct selection is then immediately set properly in the updateEditingState() call
+      // in this method. This is a hack to trigger Samsung keyboard's internal cache to clear.
+      // This prevents duplication on keyboard hide+show. See https://github.com/flutter/flutter/issues/31512
+      //
+      // We only do this if the proper selection will be restored later, eg, when mBatchCount is 0.
+      if (mBatchCount == 0) {
+        mImm.updateSelection(mFlutterView, -1, -1, -1, -1);
+      }
     }
 
     updateEditingState();
     return result;
   }
 
+  // Detect if the keyboard is a Samsung keyboard, where we apply Samsung-specific hacks to
+  // fix critical bugs that make the keyboard otherwise unusable. See finishComposingText() for
+  // more details.
+  @SuppressLint("NewApi") // New API guard is inline, the linter can't see it.
+  @SuppressWarnings("deprecation")
+  private boolean isSamsung() {
+    InputMethodSubtype subtype = mImm.getCurrentInputMethodSubtype();
+    // Impacted devices all shipped with Android Lollipop or newer.
+    if (subtype == null
+        || Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP
+        || !Build.MANUFACTURER.equals("samsung")) {
+      return false;
+    }
+    String keyboardName =
+        Settings.Secure.getString(
+            mFlutterView.getContext().getContentResolver(), Settings.Secure.DEFAULT_INPUT_METHOD);
+    // The Samsung keyboard is called "com.sec.android.inputmethod/.SamsungKeypad" but look
+    // for "Samsung" just in case Samsung changes the name of the keyboard.
```

Comment:
Do they have a history of changing the name of the keyboard and retaining existing bugs? If not, I'd be as precise as possible so we don't over-apply the hacks.

Reply:
I don't think they change the name of the keyboard, but the bug does manifest in different ways across different devices/versions of the keyboard, sometimes in a regression-like manner (newer versions can have bugs fixed in older versions).

Reply:
Makes sense. Thanks.